### PR TITLE
[SDK/CLI] Add eager flow schema

### PR DESCRIPTION
**Pull request by D-W-**
This pull request includes a variety of changes to different files in the codebase. The most important changes include removing the `context` property and its setter, adding a new class method to check if the flow is a DAG flow, and modifying the `FlowSchema` class to include new fields.

Main changes:

* [`src/promptflow/promptflow/_sdk/entities/_flow.py`](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bL145-R191): Removed the `context` property and its setter, added a new class method to check if the flow is a DAG flow, and modified the `Flow` class by removing the initialization of the `_context` attribute and adding the `FlowBase` class as a base class. [[1]](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bL145-R191) [[2]](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bR236-R239) [[3]](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bR20) [[4]](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bL121-L123) [[5]](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bL31-L45) [[6]](diffhunk://#diff-339752a6e9c04159a62708efed6b385bd91b016a0f0d2edeb32024fd19886a2bR95-R124)
* [`src/promptflow/promptflow/_sdk/entities/_eager_flow.py`](diffhunk://#diff-53d7ff49a8720f52ee715a6f63d5e389750e44c578a907076d435e541f0163eeR1-R32): Added a new class `EagerFlow` to represent an eager flow, with an `__init__` method and a class method `_load` to load the flow from a path.
* [`src/promptflow/promptflow/_sdk/schemas/_flow.py`](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53R46-R63): Modified the `FlowSchema` class to include new fields `inputs`, `outputs`, `nodes`, and `node_variants`, and renamed the `FlowSchema` class in `FlowOutputSchema` to `BaseFlowSchema` and modified its definition. [[1]](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53R46-R63) [[2]](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53L9-R9) [[3]](diffhunk://#diff-84af29d72d79c61d55daa91de77a03235646967bcaa7c3299f33adcead97bc53L34-L41)

Other changes:

* [`src/promptflow/tests/test_configs/eager_flows/multiple_entries/entry2.py`](diffhunk://#diff-bb3c84e8ed415574d1b2ac5bd47b4566f65687f99fe5491f6f185ae2121ab17cR1-R12): Added two new functions `my_flow1` and `my_flow2` to define simple flows without using YAML.
* [`src/promptflow/promptflow/_sdk/_load_functions.py`](diffhunk://#diff-823b076b6e4e3c7b30bffee4f0ee2d8e7f32a9fc1d85854c5880aef7681ae2e4L13-R13): Removed an import statement, modified a function to accept additional keyword arguments, and changed the way a function loads a flow. [[1]](diffhunk://#diff-823b076b6e4e3c7b30bffee4f0ee2d8e7f32a9fc1d85854c5880aef7681ae2e4L13-R13) [[2]](diffhunk://#diff-823b076b6e4e3c7b30bffee4f0ee2d8e7f32a9fc1d85854c5880aef7681ae2e4R65) [[3]](diffhunk://#diff-823b076b6e4e3c7b30bffee4f0ee2d8e7f32a9fc1d85854c5880aef7681ae2e4L75-R76)
* [`src/promptflow/tests/sdk_cli_test/unittests/test_flow.py`](diffhunk://#diff-1cf60dcc128520edf2745f22a4d14440fe785f03676c53918652c5c4cfb92d63R1-R77): Added a new test class `TestRun` to test the loading of flows using the `load_flow` function, covering different scenarios such as eager flows, DAG flows, flows with environment settings, and invalid flows.# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 1]
```diff
@@ -142,37 +156,39 @@ def path(self) -> Path:
...
-    def context(self) -> FlowContext:
-        return self._context
-
-    @context.setter
-    def context(self, val):
-        if not isinstance(val, FlowContext):
-            raise UserErrorException("context must be a FlowContext object, got {type(val)} instead.")
-        self._context = val
+    @classmethod
+    def _is_dag_flow(cls, data: dict):
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:160]: @wangchao1230  shall we add a field (e.g. type) to distinguish eager flow?

File: src/promptflow/promptflow/_sdk/entities/_flow.py [hunk 2]
```diff
@@ -142,37 +156,39 @@ def path(self) -> Path:
...
             with open(flow_path, "r", encoding=DEFAULT_ENCODING) as f:
                 flow_content = f.read()
-                flow_dag = yaml.safe_load(flow_content)
+                data = yaml.safe_load(flow_content)
                 kwargs["content_hash"] = hash(flow_content)
-            return cls(code=flow_path.parent.absolute().as_posix(), dag=flow_dag, **kwargs)
-
-        raise UserErrorException("Source must be a directory or a 'flow.dag.yaml' file")
+            is_dag_flow = cls._is_dag_flow(data)
+            # TODO: schema validation and warning on unknown fields
```
Comment (D-W-) [inline src/promptflow/promptflow/_sdk/entities/_flow.py:185]: @wangchao1230 I remember we'll only print warning on unknown fields, right?

File: src/promptflow/promptflow/_sdk/_load_functions.py [hunk 3]
```diff
@@ -72,7 +73,7 @@ def load_flow(
     :return: A Flow object
     :rtype: Flow
     """
-    return ProtectedFlow.load(source)
+    return Flow.load(source, **kwargs)
```
Comment (elliotzh) [inline src/promptflow/promptflow/_sdk/_load_functions.py:76]: after some other practice I agree on this change, but wonder if this is safe as we already have some logic depending on ProtectedFlow

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 1 above)
Comment (wangchao1230) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:160]: Is the yaml schema easy to describe the mutual exclusion of path and nodes?

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 2 above)
Comment (wangchao1230) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:185]: Yes. Only for unkown fields.

File: src/promptflow/promptflow/_sdk/entities/_flow.py (hunk 1 above)
Comment (D-W-) [reply src/promptflow/promptflow/_sdk/entities/_flow.py:160]: Offline synced, it's not easy to decide what value should we put here. So we currently use "entry" to decide if it's a eager flow.

File: src/promptflow/promptflow/_sdk/schemas/_flow.py [hunk 4]
```diff
@@ -47,3 +43,21 @@ class FlowSchema(YamlFileSchema):
...
+class FlowSchema(BaseFlowSchema):
+    """Schema for flow dag."""
+
+    inputs = fields.Dict(keys=fields.Str(), values=NestedField(FlowInputSchema))
+    outputs = fields.Dict(keys=fields.Str(), values=NestedField(FlowOutputSchema))
+    nodes = fields.List(fields.Dict())
+    node_variants = fields.Dict(keys=fields.Str(), values=fields.Dict())
+
+
+class EagerFlowSchema(BaseFlowSchema):
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/schemas/_flow.py:57]: Shall we include the generated schema in the repo?

Comment (D-W-) [reply src/promptflow/promptflow/_sdk/schemas/_flow.py:57]: Put eager flow and merged flow schema in here:
image

**Review by wangchao1230 — APPROVED** [review]

**Review by elliotzh — APPROVED** [review]

**Review by brynn-code — APPROVED** [review]

**Pull request merged**